### PR TITLE
Add SotoSignerCloudFront module for signed URLs and cookies

### LINKED ISSUE #613: CloudFront signed URLs and cookies
https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-setting-signed-cookie-canned-policy.html

### PR DESCRIPTION
**Pull request by sebsto**
Adds a new `SotoSignerCloudFront` module that generates CloudFront signed URLs and signed cookies using RSA PKCS#1 v1.5 signatures. This is a first-class implementation of the signing logic described in the [CloudFront developer guide](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-signed-urls.html).

Closes #613

## What's included

- **New module**: `SotoSignerCloudFront` (separate from `SotoSignerV4` — uses `_CryptoExtras` for RSA)
- **Canned policy**: signed URLs and cookies with expiration
- **Custom policy**: wildcard resources, IP restrictions, optional start times
- **SHA-1 and SHA-256**: both hash algorithms supported (`Hash-Algorithm` param appended for SHA-256)
- **Sendable**: safe across concurrency domains, Swift 6 strict concurrency clean
- **Cross-platform**: macOS + Linux, verified on `swift:amazonlinux2023`

## API

```swift
let signer = try CloudFrontSigner(keyPairId: "K2JCJMDEHXQW5F", privateKey: pemString)

// Canned policy
let url = try signer.signedURL(url: "https://d111.cloudfront.net/video.mp4", expires: .hours(1))

// Custom policy with wildcard + IP restriction
let policy = CloudFrontSigner.CustomPolicy(
    resource: "https://d111.cloudfront.net/premium/*",
    expires: .hours(24),
    ipAddress: "192.0.2.0/24"
)
let url = try signer.signedURL(url: "https://d111.cloudfront.net/premium/movie.mp4", policy: policy)

// Signed cookies
let cookies = try signer.signedCookies(url: "https://d111.cloudfront.net/file.mp4", expires: .hours(2))
```

## Testing

- 33 unit tests covering all public API methods, edge cases, and error conditions
- Verified on macOS (local) and Linux (`docker run swift:amazonlinux2023`)
- **End-to-end tested** against a real CloudFront distribution — all 5 scenarios pass (canned SHA-1, canned SHA-256, custom wildcard, signed cookies, expired URL rejection)

## Documentation

Includes Swift DocC documentation:
- Module overview with usage examples
- API reference for `CloudFrontSigner`
- **End-to-end testing guide** (`SotoCore.docc/SotoSignerCloudFront/Articles/EndToEndTesting.md`) with copy-pasteable AWS CLI commands to set up a test distribution, generate signed URLs, verify with curl, and clean up all resources

### CODE REVIEW

File: Sources/SotoSignerCloudFront/CloudFrontSigner.swift [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+            throw CloudFrontSignerError.invalidPrivateKey
+        }
+    }
+
+    /// Initialize a CloudFront signer with a key pair ID and PEM-encoded private key (Data).
+    /// - Parameters:
+    ///   - keyPairId: The CloudFront key pair ID (e.g., "K2JCJMDEHXQW5F")
+    ///   - privateKey: PEM-encoded RSA private key as Data
+    ///   - hashAlgorithm: Hash algorithm to use (defaults to `.sha1`)
+    public init(keyPairId: String, privateKey: Data, hashAlgorithm: HashAlgorithm = .sha1) throws {
```
Comment (adam-fowler) [inline Sources/SotoSignerCloudFront/CloudFrontSigner.swift]: Is this function necessary?
It would make more sense if it accepted a DER encoded key

File: Sources/SotoSignerCloudFront/CloudFrontSigner.swift [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            ipAddress: String? = nil
+        ) {
+            self.resource = resource
+            self.expires = expires
+            self.activeFrom = activeFrom
+            self.ipAddress = ipAddress
+        }
+    }
+
+    /// The three cookie values needed for CloudFront signed cookies
```
Comment (adam-fowler) [inline Sources/SotoSignerCloudFront/CloudFrontSigner.swift]: three or four

File: Sources/SotoSignerCloudFront/CloudFrontSigner+SignedCookies.swift [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            keyPairId: keyPairId
+        )
+    }
+
+    /// Generate signed cookies using a custom policy.
+    /// - Parameters:
+    ///   - policy: Custom policy specifying access conditions
+    ///   - date: Date that cookies are valid from, defaults to now
+    /// - Returns: The cookie values to set in `Set-Cookie` headers
+    public func signedCookies(policy: CustomPolicy, date: Date = Date()) throws -> SignedCookies {
```
Comment (adam-fowler) [inline Sources/SotoSignerCloudFront/CloudFrontSigner+SignedCookies.swift]: Any reason this doesn't return just a string containing the cookies, instead of expecting the user to create the cookie sting. Or alternatively have `SignedCookies` include a computed variable that is the cookie header string.

File: Sources/SotoSignerCloudFront/CloudFrontSigner.swift [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    let hashAlgorithm: HashAlgorithm
+
+    // MARK: - Initializers
+
+    /// Initialize a CloudFront signer with a key pair ID and PEM-encoded private key (String).
+    /// - Parameters:
+    ///   - keyPairId: The CloudFront key pair ID (e.g., "K2JCJMDEHXQW5F")
+    ///   - privateKey: PEM-encoded RSA private key as a String
+    ///   - hashAlgorithm: Hash algorithm to use (defaults to `.sha1`)
+    public init(keyPairId: String, privateKey: String, hashAlgorithm: HashAlgorithm = .sha1) throws {
```
Comment (adam-fowler) [inline Sources/SotoSignerCloudFront/CloudFrontSigner.swift]:
```suggestion
    public init(keyPairId: String, privateKeyPEM: String, hashAlgorithm: HashAlgorithm = .sha1) throws {
```

File: Sources/SotoSignerCloudFront/CloudFrontSigner+Internal.swift [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+extension CloudFrontSigner {
+
+    /// Apply CloudFront's URL-safe base64 encoding.
+    ///
+    /// Standard base64 is produced first, then the following replacements are applied:
+    /// - `+` → `-`
+    /// - `=` → `_`
+    /// - `/` → `~`
+    static func cloudFrontBase64Encode(_ data: Data) -> String {
```
Comment (adam-fowler) [inline Sources/SotoSignerCloudFront/CloudFrontSigner+Internal.swift:38]: In Swift 6.3 I think you can use `base64 = data.base64EncodedString(options: . base64URLAlphabet)`

File: Sources/SotoSignerCloudFront/CloudFrontSigner+SignedURL.swift [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        return signedURL
+    }
+
+    /// Generate a signed URL using a custom policy.
+    /// - Parameters:
+    ///   - url: The CloudFront URL to sign
+    ///   - policy: Custom policy specifying access conditions
+    ///   - date: Date that URL is valid from, defaults to now
+    /// - Returns: The complete signed URL
+    public func signedURL(url: String, policy: CustomPolicy, date: Date = Date()) throws -> String {
```
Comment (adam-fowler) [inline Sources/SotoSignerCloudFront/CloudFrontSigner+SignedURL.swift]: Would it be possible to combine the two `signedURL` functions and have a policy enum (probably best to make it a struct).
```swift
enum Policy {
    case custom(CustomPolicy)
    case canned(expires: Date)
}
```

Similarly do the same for Cookies

File: Sources/SotoSignerCloudFront/CloudFrontSignerError.swift [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of Soto project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+/// Errors that can occur during CloudFront signing operations
+public enum CloudFrontSignerError: Error, Sendable {
```
Comment (adam-fowler) [inline Sources/SotoSignerCloudFront/CloudFrontSignerError.swift]: Please change this to a struct

File: Sources/SotoSignerCloudFront/CloudFrontSigner.swift (hunk 2 above)
Comment (sebsto) [reply Sources/SotoSignerCloudFront/CloudFrontSigner.swift]: Fixed — removed the incorrect count. The comment now just says "The cookie values needed for CloudFront signed cookies" since the actual count depends on canned vs custom policy.

Commit: https://github.com/soto-project/soto-core/pull/696/commits/ab9c8fe5c252b67e507d7ed1e5c849e4dcc045fd

File: Sources/SotoSignerCloudFront/CloudFrontSigner.swift (hunk 1 above)
Comment (sebsto) [reply Sources/SotoSignerCloudFront/CloudFrontSigner.swift]: Done — changed the Data initializer to accept DER-encoded keys (`privateKeyDER:`) instead. The String initializer is renamed to `privateKeyPEM:`. Both are kept since PEM is what most users will have from AWS/openssl, and DER covers binary key material from Keychain etc.

Commit: https://github.com/soto-project/soto-core/pull/696/commits/71fd6418

File: Sources/SotoSignerCloudFront/CloudFrontSigner+SignedCookies.swift (hunk 3 above)
Comment (sebsto) [reply Sources/SotoSignerCloudFront/CloudFrontSigner+SignedCookies.swift]: Added a `headerValues` computed property on `SignedCookies` that returns an array of ready-to-use `Set-Cookie` header strings. The individual fields are still available for cases where users need direct access.

Commit: https://github.com/soto-project/soto-core/pull/696/commits/3b0cbb1d

File: Sources/SotoSignerCloudFront/CloudFrontSigner.swift (hunk 4 above)
Comment (sebsto) [reply Sources/SotoSignerCloudFront/CloudFrontSigner.swift]: Done — renamed to `privateKeyPEM:`.

Commit: https://github.com/soto-project/soto-core/pull/696/commits/71fd6418

File: Sources/SotoSignerCloudFront/CloudFrontSigner+Internal.swift (hunk 5 above)
Comment (sebsto) [reply Sources/SotoSignerCloudFront/CloudFrontSigner+Internal.swift:38]: Investigated this — CloudFront's encoding is not standard base64url (RFC 4648). The substitutions are `+ → -`, `/ → ~`, `= → _`, whereas base64url uses `+ → -`, `/ → _`, and strips padding. The `~` for `/` and `_` for `=` are CloudFront-specific, so `.base64URLAlphabet` can't be used here. Added a doc comment explaining this.

Commit: https://github.com/soto-project/soto-core/pull/696/commits/325d56cc

File: Sources/SotoSignerCloudFront/CloudFrontSigner+SignedURL.swift (hunk 6 above)
Comment (sebsto) [reply Sources/SotoSignerCloudFront/CloudFrontSigner+SignedURL.swift]: Done — unified both `signedURL` methods (and both `signedCookies` methods) into single methods that accept a `Policy` enum:

```swift
public enum Policy: Sendable {
    case canned(expires: TimeAmount)
    case custom(CustomPolicy)
}
```

Usage: `signer.signedURL(url: url, policy: .canned(expires: .hours(1)))`

Commit: https://github.com/soto-project/soto-core/pull/696/commits/74c8e26f

File: Sources/SotoSignerCloudFront/CloudFrontSignerError.swift (hunk 7 above)
Comment (sebsto) [reply Sources/SotoSignerCloudFront/CloudFrontSignerError.swift]: Done — converted to a struct with a private enum code and static properties for better ABI stability and extensibility.

Commit: https://github.com/soto-project/soto-core/pull/696/commits/33cec23a

File: Sources/SotoSignerCloudFront/CloudFrontSigner+Internal.swift (hunk 5 above)
Comment (sebsto) [reply Sources/SotoSignerCloudFront/CloudFrontSigner+Internal.swift:38]: Update: I verified this more carefully. `.base64URLAlphabet` (available macOS 26.4+) *could* be used as a starting point — it maps `+` → `-` and `/` → `_`, keeping padding `=`. You'd then need two more replacements (`_` → `~`, `=` → `_`) to get CloudFront's encoding. It doesn't simplify the code (still needs manual replacements) and requires a newer deployment target, so I kept the current approach but updated the doc comment to explain this accurately.

File: Sources/SotoSignerCloudFront/CloudFrontSigner.swift [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+            throw CloudFrontSignerError.invalidPrivateKey
+        }
+    }
+
+    /// Initialize a CloudFront signer with a key pair ID and DER-encoded private key.
+    /// - Parameters:
+    ///   - keyPairId: The CloudFront key pair ID (e.g., "K2JCJMDEHXQW5F")
+    ///   - privateKeyDER: DER-encoded RSA private key
+    ///   - hashAlgorithm: Hash algorithm to use (defaults to `.sha1`)
+    public init(keyPairId: String, privateKeyDER: Data, hashAlgorithm: HashAlgorithm = .sha1) throws {
```
Comment (adam-fowler) [inline Sources/SotoSignerCloudFront/CloudFrontSigner.swift]:
```suggestion
    public init<Bytes: DataProtocol>(keyPairId: String, privateKeyDER: Bytes, hashAlgorithm: HashAlgorithm = .sha1) throws {
```

File: Sources/SotoSignerCloudFront/CloudFrontSigner+Internal.swift (hunk 5 above)
Comment (adam-fowler) [reply Sources/SotoSignerCloudFront/CloudFrontSigner+Internal.swift:38]: > CloudFront's encoding is not standard base64url

Typical

File: Sources/SotoSignerCloudFront/CloudFrontSigner.swift [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        ) {
+            self.resource = resource
+            self.expires = expires
+            self.activeFrom = activeFrom
+            self.ipAddress = ipAddress
+        }
+    }
+
+    /// The signing policy to use for URL or cookie generation.
+    public enum Policy: Sendable {
```
Comment (adam-fowler) [inline Sources/SotoSignerCloudFront/CloudFrontSigner.swift]: Can we make this a struct

**Review by adam-fowler — COMMENTED** [review]
A couple of minor changes, but otherwise we are good

File: Sources/SotoSignerCloudFront/CloudFrontSigner.swift (hunk 8 above)
Comment (sebsto) [reply Sources/SotoSignerCloudFront/CloudFrontSigner.swift]: Done — changed to `<Bytes: DataProtocol>` generic.

Commit: d93e3d9e

File: Sources/SotoSignerCloudFront/CloudFrontSigner.swift (hunk 9 above)
Comment (sebsto) [reply Sources/SotoSignerCloudFront/CloudFrontSigner.swift]: Done — converted `Policy` from an enum to a struct with static factory methods (`.canned(expires:)` and `.custom(_:)`). Call-site syntax is unchanged; internal dispatch uses a private `_Policy` enum.

Commit: dd2446b7

**Comment by sebsto** [comment]
All good.

**Review by adam-fowler — APPROVED** [review]
